cmd/upgrade.go

```diff
@@ -147,23 +151,26 @@ func (uc *upgradeCmd) loadCluster(cmd *cobra.Command) error {
 		return errors.Wrap(err, "error parsing the api model")
 	}
 
-	if uc.containerService.Location == "" {
-		uc.containerService.Location = uc.location
-	} else if uc.containerService.Location != uc.location {
-		return errors.New("--location does not match api model location")
+	templatePath := path.Join(uc.deploymentDirectory, "azuredeploy.json")
+	armTemplateHandle, err := os.Open(templatePath)
+	if err != nil {
+		return errors.Wrap(err, "error reading ARM file")
+	}
+	defer armTemplateHandle.Close()
+	err = uc.validateCurrentLocalState(armTemplateHandle)
+	if err != nil {
+		return errors.Wrap(err, "error validating the api model")
 	}
+	return nil
+}
 
+func (uc *upgradeCmd) validateTargetVersion() error {
 	// Get available upgrades for container service.
 	orchestratorInfo, err := api.GetOrchestratorVersionProfile(uc.containerService.Properties.OrchestratorProfile, uc.containerService.Properties.HasWindows())
 	if err != nil {
 		return errors.Wrap(err, "error getting list of available upgrades")
 	}
 
-	// Add the current version to account for failed upgrades.
-	orchestratorInfo.Upgrades = append(orchestratorInfo.Upgrades, &api.OrchestratorProfile{
```

Comment:
the primary purpose of this was to allow upgrade operation retries on previously failed upgrades. Does this change mean it is not possible to retry a failed upgrade anymore without passing in the --force flag? we need to update the upgrade documentation to reflect this if that's the case.

I have a slight preference for not requiring the --force flag for retries. It might break CI for some users and/or encourage the use of the --force flag by default which is dangerous since it skips validation all together...

Reply:
The behaviour of the cmd validation is to validate against local state when it comes to version.
the behaviour of the upgrade process is unchanged when you don't pass --force, if the validation of the cmd goes through.
Which means that as long as your local apimodel passes the cmd validation, you get the same behaviour as before (skip upgraded nodes, upgrade nodes that are behind)

Reply:
I spent a good amount of time writing tests around operations/upgradecluster.go to not break existing behaviour.